File: Python/emg.py
import signal_utilities as su

#Simple filter class assumes no low-frequency noise and data centred at 0. More complex filteres inherit from this.
class EMG_filter_basic(object):
    def __init__(self, sample_frequency = 200, range_ = 0.1, reference_available = False):
        self.log = []                                          #used to store filtered data
        self.reference_available = reference_available
        self.MA = su.Moving_Average(length = sample_frequency * range_, return_int = True)

        if reference_available == True:
            self.LPF_data = su.LPF(cutoff_frequency = 150, sample_frequency = sample_frequency)
            self.LPF_reference = su.LPF(cutoff_frequency = 150, sample_frequency = sample_frequency)

    #input filtered data
    def log_data(self, to_log):
        self.log.insert(0, to_log)

    #pop and return either one data point or a set of data points
    def pop_log(self, num_to_pop = 1):
        if len(self.log) < 1 or num_to_pop > len(self.log):
            return
        elif num_to_pop <= 1:
            return self.log.pop()
        elif num_to_pop > 1 and num_to_pop <= len(self.log):
            to_return = self.log[-num_to_pop:]
            self.log = self.log[:-num_to_pop]
            return to_return

    #rectifies data, assuming low frequency noise already removed (reference point of 0)
    def rectify(self, to_rectify):
        return abs(to_rectify)

    #this function is called to input raw data. It returns a filtered value if it has enough samples, and also logs it
    def filter(self, data, reference_data = 0):
        if not self.reference_available:
            filtered_value = self.MA.get_movingAvg(self.rectify(data))
            self.log_data(filtered_value)
            return filtered_value

        else:
            clean_data = self.LPF_data.filter(data)
            clean_reference = self.LPF_reference.filter(reference_data)
            filtered_value = self.MA.get_movingAvg(self.rectify(clean_data - clean_reference))
            self.log_data(filtered_value)
            #print "Reference used"
            return filtered_value

#extension of EMG_filter_basic that has tools to remove low-frequency noise and normalize the data
class EMG_filter(EMG_filter_basic):

    def __init__(self, sample_frequency = 200, range_ = 0.5, min_EMG_frequency = 25, max_EMG_frequency = 150, reference_available = False):
        EMG_filter_basic.__init__(self, sample_frequency = sample_frequency, range_ = range_, reference_available = reference_available)
        #self.reference_available = reference_available
        self.PkPk = su.PkPk(sample_frequency = sample_frequency, min_frequency = min_EMG_frequency, max_frequency = max_EMG_frequency)

    #this function is called to input raw data and return a filtered value, accounting for low-frequency noise and un-normalized data
    def filter(self, data, reference_data = 0):
        if self.reference_available == True:
            clean_data = self.LPF_data.filter(data)
            clean_reference = self.LPF_reference.filter(reference_data)
            data = clean_data - clean_reference
        
        neutral_value = self.PkPk.get_pkpk(data)['neutral']
        filtered_value = self.MA.get_movingAvg(self.rectify(data - neutral_value))
        self.log_data(filtered_value)
        return filtered_value

        #return super(EMG_filter, self).filter(data - neutral_value)

#print "End of EMG module"

File: C/emg.c
#include "emg.h"

#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>

struct EMG_ {

	Moving_Average *MA;

	EMG_OPTIONS remove_low_frequency;
	Moving_Average *For_HPF;
	//PkPk *PkPk;

	EMG_OPTIONS reference_available;
	Moving_Average *LPF_data;
	Moving_Average *LPF_reference;

};

EMG *new_EMG(	uint16_t sample_frequency_, 
				float range_, 
				uint16_t min_EMG_frequency_, 
				uint16_t max_EMG_frequency_, 
				EMG_OPTIONS remove_low_frequency_, 
				EMG_OPTIONS reference_available_) 
{
	EMG *to_return = (EMG *)malloc(sizeof(EMG));
	if (!to_return) {
		puts("Could not allocate memory for EMG");
		return NULL;
	}

	to_return->MA = new_moving_average(sample_frequency_ * range_);
	if (!to_return->MA) {
		puts("Could not allocate memory for internal data structures");
		return NULL;
	}

	to_return->remove_low_frequency = remove_low_frequency_;
	if (to_return->remove_low_frequency == HIGH_PASS_FILTER_ON) {
		to_return->For_HPF = new_moving_average(sample_frequency_ * 2 / min_EMG_frequency_);
		if(!to_return->For_HPF){
			puts("Could not allocate memory for internal data structures");
			return NULL;
		}
		/*
		to_return->PkPk = new_pkpk(sample_frequency_, min_EMG_frequency_, max_EMG_frequency_);
		if (!to_return->PkPk) {
			puts("Could not allocate memory for internal data structures");
			return NULL;
		}
		*/
	} else {
		to_return->For_HPF = NULL;
		//to_return->PkPk = NULL;
	}

	to_return->reference_available = reference_available_;
	if (to_return->reference_available == REFERENCE_AVAILABLE) {
		uint16_t length = (0.125 * sample_frequency_ / max_EMG_frequency_ >= 1)?0.125 * sample_frequency_ / max_EMG_frequency_:1;
		to_return->LPF_data = new_moving_average(length);
		to_return->LPF_reference = new_moving_average(length);
		if (!to_return->LPF_data || !to_return->LPF_reference) {
			puts("Could not allocate memory for internal data structures");
			return NULL;
		}
	} else {
		//to_return->LPF_data = NULL;
		uint16_t length = (0.125 * sample_frequency_ / max_EMG_frequency_ >= 1)?0.125 * sample_frequency_ / max_EMG_frequency_:1;
		to_return->LPF_data = new_moving_average(length);

		if (!to_return->LPF_data) {
			puts("Could not allocate memory for internal data structures");
			return NULL;
		}

		to_return->LPF_reference = NULL;
	}

	return to_return;
}

void free_EMG(EMG *self) 
{
	puts("Freeing EMG");
	free_moving_average(self->MA);
	self->MA = NULL;
	if (self->For_HPF) {
		free_moving_average(self->For_HPF);
		self->For_HPF = NULL;
	}
	/*
	if (self->PkPk) {
		free_pkpk(self->PkPk);
		self->PkPk = NULL;
	}
	*/
	if (self->LPF_data) {
		free_moving_average(self->LPF_data);
		self->LPF_data = NULL;
	}
	if (self->LPF_reference) {
		free_moving_average(self->LPF_reference);
		self->LPF_reference = NULL;
	}
	free(self);
	self = NULL;
}

int filter_EMG(	EMG *self, 
				int data) 
{
	if (self->reference_available != REFERENCE_AVAILABLE) {
		data = get_moving_average(self->LPF_data, data);
	}

	if (self->remove_low_frequency == HIGH_PASS_FILTER_ON) {
		//int neutral_value = unpack_data(get_pkpk(self->PkPk, data), PKPK_NEUTRAL);
		int neutral_value = get_moving_average(self->For_HPF, data);
		return get_moving_average(self->MA, abs(data - neutral_value));
	} else {
		return get_moving_average(self->MA, abs(data));
	}
}

int filter_EMG_r(	EMG *self, 
					int data, 
					int reference_data) 
{
	if (self->reference_available != REFERENCE_AVAILABLE) {
		return -1;
	}

	int clean_data = get_moving_average(self->LPF_data, data);
	int clean_reference = get_moving_average(self->LPF_reference, reference_data);
	data = clean_data - reference_data;

	return filter_EMG(self, data);
}